text/3493-precise-pre-release-cargo-update.md

```diff
@@ -0,0 +1,241 @@
+`x.y.z-pre.0` is considered compatible with `a.b.c` when requested `--precise`ly if `x.y.z` is semver compatible with `a.b.c` and `a.b.c` `!=` `x.y.z`.
+
+Consider a `Cargo.toml` with this `[dependencies]` section
+
+```
+[dependencies]
+example = "1.0.0"
+```
+
+It is possible to update to `1.2.0-pre.0` because `1.2.0` is semver compatible with `1.0.0`
+
+```
+> cargo update -p example --precise 1.2.0-pre.0
+    Updating crates.io index
+    Updating example v1.0.0 -> v1.2.0-pre.0
+```
+
+It is not possible to update to `2.0.0-pre.0` because `2.0.0` is not semver compatible with `1.0.0`
+
+```
+> cargo update -p example --precise 2.0.0-pre.0
+    Updating crates.io index
+error: failed to select a version for the requirement `example = "^1"`
+candidate versions found which didn't match: 2.0.0-pre.0
+location searched: crates.io index
+required by package `tmp-oyyzsf v0.1.0 (/home/ethan/.cache/cargo-temp/tmp-OYyZsF)`
+```
+
+| Cargo.toml     | Cargo.lock    | Desired       | Selectable w/o `--precise` | Selectable w/ `--precise`  |
+| ---------------| --------------| --------------| -------------------------- | ---------------------------|
+| `^1.0.0`       | `1.0.0`       | `1.0.0-pre.0` | ❌                         | ❌                         |
+| `^1.0.0`       | `1.0.0`       | `1.2.0`       | ✅                         | ✅                         |
+| `^1.0.0`       | `1.0.0`       | `1.2.0-pre.0` | ❌                         | ✅                         |
+| `^1.0.0`       | `1.2.0-pre.0` | `1.2.0-pre.1` | ❌                         | ✅                         |
+| `^1.2.0-pre.0` | `1.2.0-pre.0` | `1.2.0-pre.1` | ✅¹                        | ✅                         |
+| `^1.2.0-pre.0` | `1.2.0-pre.0` | `1.3.0`       | ✅¹                        | ✅                         |
+
+✅: Will upgrade
+
+❌: Will not upgrade
+
+¹This behaviour is considered by some to be undesirable and may change as proposed in [RFC: Precise Pre-release Deps](https://github.com/rust-lang/rfcs/pull/3263).
+This RFC preserves this behaviour to remain backwards compatible.
+Since this RFC is concerned with the behaviour of `cargo update --precise` changes to bare `cargo update` made in future RFCs should have no impact on this proposal.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Version requirements
+
+Version requirement operator semantics will change to encompass pre-release versions, compared to before where the presence of pre-release would change matching modes.
+This will also better align with the mathematical properties associated with some of the operators (see the closed [RFC 3266](https://github.com/rust-lang/rfcs/pull/3266)).
+
+So before,
+```
+1.2.3  -> ^1.2.3 -> >=1.2.3, <2.0.0 (with implicit holes excluding pre-release versions)
+```
+would become
+```
+1.2.3  -> ^1.2.3 -> >=1.2.3, <2.0.0-0
+```
```

Comment:
For packages that use `<` (e.g. encompassing multiple semver major versions), this will be a breaking change because they could then start matching pre-releases for the next major version.  We might need to special case `<`.

Reply:
We talked about this in today's cargo meeting.

For `<` we likely want to keep the property of "doesn't match pre-release unless a pre-release is present" or in other words, implicitly add a `-0` to `<` version requirements.

That said, this leaves a corner case for `>=0.14-0,<0.14.0`.
- Do we make one comparator dependent on another and detect the initial pre-release?
  - Distant enough effects that this makes it hard to predict
- Do nothing and accept the corner case

Reply:
I forgot to mention that some of the details here can likely be left in Unresolved Questions to be figured out during the implementation and we can affirm those decisions during stabilization.

Reply:
Thanks Ed. I've added https://github.com/rust-lang/rfcs/pull/3493/commits/ec033c91da1e36cf6c422f10043df95160c1e85b to Unresolved Questions. Hopefully this sufficiently captures the details of the corner case.

Reply:
Thanks!